content/blog/2019-09-25-bril-to-llvm-ocaml.md

```diff
@@ -0,0 +1,151 @@
+
+```ocaml
+match op with 
+| Id {dest;_}
+| Const {dest; _}
+| Add {dest;_}
+... -> Some dest
+| _ -> None
+```
+
+In typescript I could do something like:
+
+```typescript
+if (op.dest) {
+    return op.dest;
+} else {
+    return null;
+}
+```
+
+The above code does not need to know which operation it is operating on, only if the operation contains a specific piece of data.
+
+One of the goals that I came up with for the representation was that I should have a representation that allowed me to match on an operation based on the data it contained or match on a specific operation. Furthermore, if I matched on the data in an operation, this should  statically limit the kinds of operations that I could match on. For example, if the case of a match statement I am in tells me that I have a `dest` field, OCaml should complain if I try to match the opcode of that operation with `Br`, since `Br` does not have a `dest` field.
+
+The way I did this was by combining GADT's with polymorphic variants to create what I called constrained extensible records. The top level record type for an operation was:
+
+```ocaml
+type 'a operation = {op: 'a opcode; ex: 'a op_ex}
+```
+
+Every operation has an opcode field. Furthermore, that opcode encodes information about the type of data held in the `ex` field of the `operation` record. An opcode is a GADT that can only be constructed using a Constructor representing one of the Bril opcodes:
+
+```ocaml
+type _ opcode =
+  | Jmp : [`Jmp] opcode
+  | Br : [`Br] opcode
+  | Add : [`Add] opcode
+  ...
+```
+
+This is where the polymorphic variants come in. The `'a` in the `operation` record is a polymorphic variant representing the operation. The polymorphic variant constrains the type we can put into the `ex` field of the `operation` record. The `op_ex` type looks like:
+
+```ocaml
+type _ op_ex =
+  | Effect_op : 'a effect_op -> 'a op_ex
+  | Mutation_op : 'a mutation_op -> 'a op_ex
+  | Nop_op : [`Nop] op_ex
+  ...
+```
+
+The basic idea was that the GADT parameterized over polymorphic variants representing the opcodes would constrain the type of data that could be in the `ex` field of an operation. This kind of pattern continues down the type hierarchy and took a while to get correct. Even though this is the representation I ended up doing this project in, the reason I called it an unsuccessful representation is that in hindsight a similar effect could have been achieved by simply defining a hierarchy of plain old variants with different data structures. I thought I would be saving myself unnecessary match statements by incorporating GADTs to constrain the data but ultimately I don't think I saved myself any match cases and had to spend a lot of time wrestling with the OCaml type system.
```

Comment:
This is a really interesting set of points! I appreciate your humility in concluding that this approach, while interesting and fun to think about, had its drawbacks.

Reply:
Thanks! I was going to leave it out originally but then someone encouraged me leave it in because they felt like talking about failure is actually really useful because it leads to less people making the same mistakes. I thought that was a very insightful point.